File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:

        dp = [0] + [amount + 1 for i in range(amount)]

        #Time Complexity : O(N^2)
        #Space Complexity : O(N)

        for i in range(1, amount+1):
            for coin in coins:
                if i < coin:
                    continue
                dp[i] = min(dp[i], dp[i-coin] + 1)                    

        return dp[-1] if dp[-1] != amount + 1 else -1                                                                       

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {

        int[] arr = new int[amount+1];
        Arrays.fill(arr, amount+1);
        arr[0] = 0;

        for(int i=1; i<amount+1; i++){
            for(int coin : coins){
                if (coin > i) continue;

                arr[i] = Math.min(arr[i], arr[i-coin] + 1);
            }
        }
        
        return arr[arr.length-1] == amount+1 ? -1 : arr[arr.length-1];
    }
}